internal/cmd/precise-code-intel-tester/upload.go

```diff
@@ -112,11 +115,55 @@ func makeProcessedSignals(revsByRepo map[string][]string) map[string]map[string]
 	return processedSignals
 }
 
+// refreshedSignals returns a map of error channels for each repository.
+func makeRefreshedSignals(revsByRepo map[string][]string) map[string]chan error {
+	refreshedSignals := map[string]chan error{}
+	for repo, revs := range revsByRepo {
+		// Each channel may receive two values for each revision: a sentinel
+		// error indicating that the repo is out of date, and a nil error when
+		// the repository's commit graph has been refreshed. At any point, the
+		// channel can receive an error value and be closed.
+		refreshedSignals[repo] = make(chan error, len(revs)*2)
+	}
+
+	return refreshedSignals
+}
+
 // watchStateChanges maintains a list of uploaded but nonterminal upload records. This function
 // polls the API and signals the worker when their upload has been processed. If an upload fails
 // to process, the error will be sent to the worker.
-func watchStateChanges(ctx context.Context, uploaded chan Upload, processedSignals map[string]map[string]chan error) {
+func watchStateChanges(
+	ctx context.Context,
+	uploaded chan Upload,
+	processedSignals map[string]map[string]chan error,
+	refreshedSignals map[string]chan error,
+) {
+	send := func(err error) {
+		// Send err to everybody and exit
+		for name, revs := range processedSignals {
+			for rev := range revs {
+				ch := processedSignals[name][rev]
```

Comment:
Nitpick, but I think this would be easier to read:

```suggestion
			for rev, ch  := range revs {
```